# **Laboratorio 8**
- Joaquín Campos - 22155
- Sofía García - 22210
- Julio García Salas - 22076

## **Inciso 1**

In [0]:
# INCISO 1 — con el nuevo maestro: /Volumes/workspace/default/lab8/master_unico.csv
# Versión compatible con Serverless (sin cache/persist)

from pyspark.sql import functions as F
import pandas as pd

# -----------------------------------------------------------
# 0) Paths y verificación
# -----------------------------------------------------------
PARENT = "/Volumes/workspace/default/lab8"
PATH   = "/Volumes/workspace/default/lab8/master_unico.csv"

display(dbutils.fs.ls(PARENT))

print("=== master_unico.csv (inicio) ===")
try:
    raw = dbutils.fs.head(PATH, 2048)
    print(raw[:1000])
except Exception as e:
    print("[WARN] No se pudo leer cabecera con dbutils.fs.head:", e)

# -----------------------------------------------------------
# 1) Carga robusta: pandas (utf-8-sig) -> Spark, con fallback a spark.read.csv
# -----------------------------------------------------------
df_all = None
try:
    pdf = pd.read_csv(PATH, dtype=str, encoding="utf-8-sig", engine="python")
    pdf.columns = [c.strip() for c in pdf.columns]
    df_all = spark.createDataFrame(pdf)
    print("[OK] Cargado con pandas -> Spark")
except Exception as e:
    print("[WARN] pandas.read_csv falló, probando spark.read.csv. Error:", e)
    df_all = (
        spark.read.format("csv")
        .option("header", True)
        .option("multiLine", True)
        .option("escape", '"')
        .load(PATH)
        .toDF(*[c.strip() for c in spark.read.csv(PATH, header=True).columns])
    )
    print("[OK] Cargado con spark.read.csv")

# 🚫 NO usar cache()/persist() en Serverless (no soportado)
# Materializamos una vez con count() y seguimos
total_rows = df_all.count()
print(f"\nTotal filas en master_unico: {total_rows:,}")
print(f"Total columnas: {len(df_all.columns)}")
print("Columnas (primeras 15):", df_all.columns[:15])

# Helper is_empty seguro (evita .isEmpty() si no está disponible)
def is_empty(sdf):
    try:
        return sdf.isEmpty()
    except Exception:
        return sdf.limit(1).count() == 0

# -----------------------------------------------------------
# 2) Separación por “tablas”
# -----------------------------------------------------------
if "source_table" in df_all.columns:
    hechos_df    = df_all.filter(F.col("source_table") == "HECHOS")
    vehiculos_df = df_all.filter(F.col("source_table") == "VEHICULOS")
    personas_df  = df_all.filter(F.col("source_table") == "PERSONAS")
else:
    hechos_df = df_all.filter(F.col("tipo_eve_norm").isNotNull())
    vehiculos_df = df_all.filter(
        F.col("tipo_veh_norm").isNotNull() |
        F.col("color_std").isNotNull() |
        F.col("sexo_pil_norm").isNotNull() |
        F.col("edad_pil_num").isNotNull()
    )
    personas_df = df_all.filter(
        F.col("sexo_per_norm").isNotNull() |
        F.col("edad_per_num").isNotNull() |
        F.col("is_fallecido").isNotNull() |
        F.col("is_lesionado").isNotNull()
    )

print("\n=== Conteos por 'tabla' ===")
if "source_table" in df_all.columns:
    df_all.groupBy("source_table").count().orderBy("source_table").show(truncate=False)
else:
    print(f"Hechos: {hechos_df.count():,} | Vehículos: {vehiculos_df.count():,} | Personas: {personas_df.count():,}")

# -----------------------------------------------------------
# 3) Muestras .show()
# -----------------------------------------------------------
print("\n--- Muestra HECHOS ---")
hechos_cols = [c for c in [
    "anio_norm","mes_norm","dia_norm","hora_clean","franja_horaria","dow_num","dow_label",
    "depto_code","depto_name_std","muni_code","zona_ciudad_num","tipo_eve_norm","anio_mes"
] if c in hechos_df.columns]
hechos_df.select(*hechos_cols).show(10, truncate=False)

print("\n--- Muestra VEHICULOS ---")
veh_cols = [c for c in [
    "anio_norm","mes_norm","hora_clean","depto_code","tipo_eve_norm",
    "tipo_veh_norm","color_std","sexo_pil_norm","edad_pil_num","marca_veh","modelo_veh"
] if c in vehiculos_df.columns]
vehiculos_df.select(*veh_cols).show(10, truncate=False)

print("\n--- Muestra PERSONAS ---")
per_cols = [c for c in [
    "anio_norm","mes_norm","hora_clean","depto_code","tipo_eve_norm",
    "sexo_per_norm","edad_per_num","is_fallecido","is_lesionado"
] if c in personas_df.columns]
personas_df.select(*per_cols).show(10, truncate=False)

# -----------------------------------------------------------
# 4) Describe & Summary de numéricas (sin cache)
# -----------------------------------------------------------
def safe_describe_numeric(sdf, cols, title):
    cols_exist = [c for c in cols if c in sdf.columns]
    if not cols_exist or is_empty(sdf):
        print(f"\n[INFO] Sin filas o columnas para {title}")
        return
    casted = [F.col(c).cast("double").alias(c) for c in cols_exist]
    print(f"\n--- DESCRIBE: {title} ---")
    sdf.select(*casted).describe().show(truncate=False)
    print(f"\n--- SUMMARY: {title} ---")
    sdf.select(*casted).summary().show(truncate=False)

safe_describe_numeric(
    hechos_df,
    ["anio_norm","mes_norm","dia_norm","hora_clean","dow_num","depto_code","muni_code","zona_ciudad_num"],
    "HECHOS (tiempo y ubicación)"
)
safe_describe_numeric(
    vehiculos_df,
    ["anio_norm","depto_code","edad_pil_num"],
    "VEHICULOS (características/conductor)"
)
safe_describe_numeric(
    personas_df,
    ["anio_norm","depto_code","edad_per_num","is_fallecido","is_lesionado"],
    "PERSONAS (atributos/estado)"
)

# -----------------------------------------------------------
# 5) Vistas temporales
# -----------------------------------------------------------
df_all.createOrReplaceTempView("master_unico")
hechos_df.createOrReplaceTempView("hechos")
vehiculos_df.createOrReplaceTempView("vehiculos")
personas_df.createOrReplaceTempView("personas")

print("\n[OK] Vistas creadas: master_unico, hechos, vehiculos, personas (sin cache/persist)")


path name size modificationTime dbfs:/Volumes/workspace/default/lab8/_UNIFICADO.csv _UNIFICADO.csv 10395125 1759426297000 dbfs:/Volumes/workspace/default/lab8/master_unico.csv master_unico.csv 29043186 1759470343000

=== master_unico.csv (inicio) ===
[Truncated to first 2048 bytes]
source_table,anio_norm,mes_norm,dia_norm,hora_clean,franja_horaria,dow_num,dow_label,depto_code,depto_name_std,muni_code,zona_ciudad_num,tipo_eve_norm,anio_mes,color_std,tipo_veh_norm,sexo_pil_norm,edad_pil_num,sexo_per_norm,edad_per_num,is_fallecido,is_lesionado,marca_veh,modelo_veh,color_veh,color_v,tipo_v,modelo_v,source_file
HECHOS,2013,1,1,3,MADRUGADA,2,MARTES,1,GUATEMALA,106,,NAN,2013-01,,,,,,,,,,,,,,,2013.csv
HECHOS,2013,1,1,5,MADRUGADA,2,MARTES,1,GUATEMALA,101,,NAN,2013-01,,,,,,,,,,,,,,,2013.csv
HECHOS,2013,1,1,3,MADRUGADA,2,MARTES,1,GUATEMALA,108,,NAN,2013-01,,,,,,,,,,,,,,,2013.csv
HECHOS,2013,1,1,0,MADRUGADA,2,MARTES,19,ZACAPA,1901,,NAN,2013-01,,,,,,,,,,,,,,,2013.csv
HECHOS,2013,1,1,18,NOCHE,2,MARTES,19,ZACAPA,1901,,NAN,2013-01,,,,,,,,,,,,,,,2013.csv
HECHOS,2013,1,1,0,MADRUGADA,2,MARTES,5,ESCUINTLA,502,,NAN,2013-01,,,,,,,,,,,,,,,2013.csv
HECHOS,2013,1,1,11,MAÑANA,2,MARTES,9,QUETZALTENANGO,901,,NAN,2013-01,,,,,,

# Resumen técnico — `master_unico.csv`

## 1) Estructura general
- **Filas:** 298,164  
- **Columnas:** 29  
- **Tablas unificadas (en `source_table`):**
  - **HECHOS:** 76,759
  - **VEHICULOS:** 111,988
  - **PERSONAS:** 109,417

> El esquema contiene campos normalizados para tiempo (`anio_norm`, `mes_norm`, `dia_norm`, `hora_clean`, `franja_horaria`, `dow_num`, `dow_label`), ubicación (`depto_code`, `depto_name_std`, `muni_code`, `zona_ciudad_num`), tipología (`tipo_eve_norm`, `tipo_veh_norm`, `color_std`) y atributos específicos de vehículo/persona.

---

## 2) Ejemplos de registros (muestras)
- **HECHOS:** filas de 2013 con `MADRUGADA/MAÑANA`, presencia de códigos de departamento y municipio; `tipo_eve_norm` aparece como **`NAN`** en algunas filas tempranas (indica valores faltantes/no mapeados).
- **VEHICULOS:** registros de 2021 con `tipo_eve_norm` ya estandarizado (p.ej., `COLISION`, `DERRAPE`), `tipo_veh_norm` y `color_std` numéricos/categorizados.
- **PERSONAS:** registros de 2021 con `sexo_per_norm`, `edad_per_num` y banderas `is_fallecido`/`is_lesionado`.

---

## 3) Cobertura de campos clave (calidad de datos)
**HECHOS** (base: 76,759 filas)  
- `hora_clean`: 76,732 (≈ **99.96%** cobertura)  
- `dow_num`, `depto_code`, `muni_code`: 60,617 cada uno (≈ **79%** cobertura)  
- `zona_ciudad_num`: 8,153 (≈ **10.6%** cobertura; muy escaso, usar con cuidado)  
- Rango temporal: **2013–2023**

**VEHICULOS** (base: 111,988 filas)  
- `depto_code`: 87,552 (≈ **78.2%**)  
- `edad_pil_num`: 5,382 (≈ **4.8%**; muy baja, sesga análisis por edad del piloto)  
- Rango temporal: **2013–2023**

**PERSONAS** (base: 109,417 filas)  
- `depto_code`: 87,497 (≈ **80.0%**)  
- `edad_per_num`: 84,249 (≈ **77.0%**)  
- `is_fallecido`: media **1.0** y `is_lesionado`: media **0.0** → **la tabla PERSONAS representa fallecidos** (no lesionados).  
- Rango temporal: **2013–2023**

**Implicación:** para preguntas que requieran **lesionados**, se deben derivar desde HECHOS/VEHICULOS (si existe la codificación) o incorporar una fuente adicional; la tabla PERSONAS no aporta lesionados.

---

## 4) Estadística descriptiva (resumen útil)
**HECHOS**  
- `anio_norm` medio ≈ **2018.27**, distribución homogénea 2013–2023.  
- `hora_clean` media ≈ **13.69** (sesgo leve hacia tarde—noche).  
- `dow_num` media ≈ **4.34** (cercano a **jueves**), con ~21% faltantes.

**VEHICULOS**  
- `anio_norm` medio ≈ **2018.49**.  
- `edad_pil_num` media ≈ **33.0** (desvío ~11.4) sobre **solo 4.8%** de filas (no representativo si no se filtra).

**PERSONAS**  
- `anio_norm` medio ≈ **2018.12**.  
- `edad_per_num` media ≈ **31.0** (desvío ~16.26).  
- Banderas: `is_fallecido=1` y `is_lesionado=0` en **100%** de registros → confirma naturaleza de fallecidos.

---

## 5) Observaciones de normalización/categorización
- `tipo_eve_norm`: aparecen valores **`NAN`** en HECHOS (2013), señal de filas no clasificadas; conviene **completar/estandarizar** para no perderlas en agregaciones por tipo.  
- `color_std`, `tipo_veh_norm`: están presentes en VEHICULOS; coherentes con la normalización previa (mapeos numéricos/códigos).  
- `depto_name_std`: disponible en HECHOS, útil para rotular gráficos; en VEHICULOS/PERSONAS se prioriza `depto_code` (80%± cobertura).

---






## **Inciso 2**

In [0]:
# INCISO 2 — Años disponibles por tabla y validación de coincidencia
# Requiere que existan las vistas temporales: `hechos`, `vehiculos`, `personas`
# (quedan creadas por el bloque del inciso 1 sobre `master_unico.csv`)

from pyspark.sql import functions as F

def years_count_df(view_name: str, year_col: str = "anio_norm"):
    """
    Devuelve un DataFrame con: anio, count
    - Filtra solo años válidos (exactamente 4 dígitos).
    - Convierte a entero para ordenar correctamente.
    """
    df = spark.table(view_name)
    if year_col not in df.columns:
        raise ValueError(f"La vista '{view_name}' no tiene la columna '{year_col}'.")
    base = (
        df.select(F.col(year_col).cast("string").alias("y"))
          .filter(F.col("y").rlike(r"^\d{4}$"))
          .withColumn("anio", F.col("y").cast("int"))
    )
    return base.groupBy("anio").count().orderBy("anio")

def table_totals(view_name: str, year_col: str = "anio_norm"):
    """
    Retorna (total_filas, filas_con_anio_valido, df_por_anio)
    """
    df = spark.table(view_name)
    total = df.count()
    df_years = years_count_df(view_name, year_col)
    valid = df_years.agg(F.sum("count").alias("valid")).collect()[0]["valid"] or 0
    return total, valid, df_years

# 1) Conteos por año para cada tabla
tot_h, valid_h, hechos_y = table_totals("hechos")
tot_v, valid_v, vehiculos_y = table_totals("vehiculos")
tot_p, valid_p, personas_y = table_totals("personas")

print("=== Hechos: años disponibles y conteos ===")
display(hechos_y)
print(f"Registros con año válido: {valid_h:,} / total: {tot_h:,}\n")

print("=== Vehículos: años disponibles y conteos ===")
display(vehiculos_y)
print(f"Registros con año válido: {valid_v:,} / total: {tot_v:,}\n")

print("=== Personas (fallecidos): años disponibles y conteos ===")
display(personas_y)
print(f"Registros con año válido: {valid_p:,} / total: {tot_p:,}\n")

# 2) Validación de coincidencia de años entre tablas
H = set(r["anio"] for r in hechos_y.select("anio").distinct().collect())
V = set(r["anio"] for r in vehiculos_y.select("anio").distinct().collect())
P = set(r["anio"] for r in personas_y.select("anio").distinct().collect())

interseccion = sorted(H & V & P)
union_total = sorted(H | V | P)
faltan_en_H = sorted(set(union_total) - H)
faltan_en_V = sorted(set(union_total) - V)
faltan_en_P = sorted(set(union_total) - P)

print("=== Validación de años entre tablas ===")
print(f"Intersección (en TODAS): {interseccion}")
print(f"Unión (en ALGUNA):      {union_total}")
print(f"- HECHOS:    años ausentes vs unión → {faltan_en_H}")
print(f"- VEHICULOS: años ausentes vs unión → {faltan_en_V}")
print(f"- PERSONAS:  años ausentes vs unión → {faltan_en_P}")


=== Hechos: años disponibles y conteos ===


anio count 2013 6324 2014 5651 2015 6854 2016 7964 2017 5879 2018 6395 2019 7047 2020 6350 2021 8153 2022 7924 2023 8218

Registros con año válido: 76,759 / total: 76,759

=== Vehículos: años disponibles y conteos ===


anio count 2013 6323 2014 7904 2015 9823 2016 11618 2017 8644 2018 9514 2019 10827 2020 10103 2021 12796 2022 12239 2023 12197

Registros con año válido: 111,988 / total: 111,988

=== Personas (fallecidos): años disponibles y conteos ===


anio count 2013 9060 2014 8990 2015 10397 2016 11668 2017 8625 2018 9407 2019 10664 2020 8142 2021 10544 2022 10722 2023 11198

Registros con año válido: 109,417 / total: 109,417

=== Validación de años entre tablas ===
Intersección (en TODAS): [2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]
Unión (en ALGUNA):      [2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]
- HECHOS:    años ausentes vs unión → []
- VEHICULOS: años ausentes vs unión → []
- PERSONAS:  años ausentes vs unión → []


# Inciso 2 — Años disponibles por tabla y validación

## Resumen ejecutivo
- **Cobertura temporal consistente (2013–2023)** en **las tres tablas**: HECHOS, VEHÍCULOS y PERSONAS.
- **Conteos totales**:
  - **HECHOS:** 76,759
  - **VEHÍCULOS:** 111,988
  - **PERSONAS (fallecidos):** 109,417
- **Relaciones útiles** (sobre todo el periodo):
  - ≈ **1.46 vehículos por accidente** (111,988 / 76,759).
  - ≈ **1.43 fallecidos por accidente** (109,417 / 76,759).  
    Esto sugiere múltiples víctimas por hecho en numerosos eventos.

---

## HECHOS — Años y tendencias
- Años presentes y **conteo por año**: 2013–2023, todos con registros.
- **Pico** en **2023** (8,218) y segundo pico en **2021** (8,153).
- **Bajada en 2020** (6,350), coherente con una posible reducción de movilidad durante la pandemia.
- **Patrón general**: leve crecimiento de 2013–2016, oscilación 2017–2019, caída en 2020, repunte marcado 2021–2023.

## VEHÍCULOS — Años y tendencias
- Años presentes y **conteo por año**: 2013–2023, todos con registros.
- **Máximo** en **2021** (12,796), con ligera **corrección** en 2022–2023 (12,239 y 12,197).
- El volumen supera sistemáticamente a HECHOS (≥1 vehículo por evento, con frecuentes >1), lo cual habilita el cálculo de **promedio de vehículos por accidente** por año/departamento.

## PERSONAS (FALLECIDOS) — Años y tendencias
- Años presentes y **conteo por año**: 2013–2023, todos con registros.
- **Mínimo** en **2020** (8,142) y **máximo** en **2023** (11,198).
- El **descriptor** previo indica `is_fallecido=1` e `is_lesionado=0` en la tabla, por lo que esta partición representa **solo fallecidos**. El conteo superior a HECHOS es consistente con **múltiples fallecidos por hecho** en diversos casos.

---

## Validación de coincidencia de años
- **Intersección y unión** de años entre tablas:  
  **[2013, 2014, 2015, 2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023]**
- **Ausencias por tabla vs unión:**  
  - **HECHOS:** []  
  - **VEHÍCULOS:** []  
  - **PERSONAS:** []
- **Conclusión:** cobertura temporal **perfectamente alineada**; no hay huecos entre tablas, lo cual **facilita joins** por año y comparaciones intertabla para los incisos posteriores.

---

## Implicaciones analíticas
- El alineamiento 2013–2023 permite:
  - Calcular y graficar **accidentes por año y departamento** sin ajustes de ventana temporal.
  - Estimar **vehículos por accidente** y **fallecidos por accidente** por año/departamento con consistencia.
  - Aislar el **efecto 2020** (posible pandemia) y contrastarlo con la **recuperación 2021–2023**.
- Para joins robustos en incisos 8–12, conviene usar llaves normalizadas:  
  `anio_norm`, `mes_norm`, `depto_code` (y opcionalmente `depto_name_std`) y `tipo_eve_norm`.



## **Inicso 3**

In [0]:
# INCISO 3 — Valores distintos de tipo de accidente (usando el master ya cargado)
# Se asume que ya existen las vistas temporales: master_unico, hechos, vehiculos, personas

from pyspark.sql import functions as F

# === 1) Usar HECHOS como fuente canónica de "tipo de accidente" ===
hechos = spark.table("hechos")

# Limpieza defensiva: pasar a string, recortar, upper; tratar "NAN"/"NULL"/vacíos como nulos
hechos_tipo = (
    hechos
    .withColumn("tipo_eve_norm_str", F.upper(F.trim(F.col("tipo_eve_norm").cast("string"))))
    .withColumn(
        "tipo_eve_norm_str",
        F.when(F.col("tipo_eve_norm_str").isin("", "NULL", "NAN", "NONE", "NA"), F.lit(None))
         .otherwise(F.col("tipo_eve_norm_str"))
    )
)

# === 2) Distintos y conteos globales (HECHOS) ===
distinct_tipos = (
    hechos_tipo
    .groupBy("tipo_eve_norm_str")
    .count()
    .orderBy(F.desc("count"))
)

print("=== Distintos 'tipo_eve_norm' con conteo (HECHOS) ===")
distinct_tipos.show(100, truncate=False)   # Ajustar 100 si se desea ver más

# === 3) Solo listado de valores distintos (sin conteos) ===
print("\n=== Valores distintos (solo listado) ===")
(
    hechos_tipo
    .select("tipo_eve_norm_str")
    .distinct()
    .orderBy("tipo_eve_norm_str")
).show(200, truncate=False)

# === 4) (Opcional) Aparición por año para cruzar con inciso 2 ===
por_anio = (
    hechos_tipo
    .groupBy("anio_norm", "tipo_eve_norm_str")
    .count()
    .orderBy("anio_norm", F.desc("count"))
)

print("\n=== Aparición por año (tipo_eve_norm) ===")
por_anio.show(100, truncate=False)

# === 5) Visualizaciones Databricks (usar display para tabla o gráfico) ===
display(distinct_tipos)  # gráfico de barras: eje X = tipo_eve_norm_str, eje Y = count
display(por_anio)        # gráfico de columnas apiladas o tabla por año

# === 6) (Opcional) Validar consistencia entre tablas del master ===
dist_por_tabla = (
    spark.table("master_unico")
    .withColumn("tipo_eve_norm_str", F.upper(F.trim(F.col("tipo_eve_norm").cast("string"))))
    .withColumn(
        "tipo_eve_norm_str",
        F.when(F.col("tipo_eve_norm_str").isin("", "NULL", "NAN", "NONE", "NA"), F.lit(None))
         .otherwise(F.col("tipo_eve_norm_str"))
    )
    .groupBy("source_table", "tipo_eve_norm_str")
    .count()
    .orderBy("source_table", F.desc("count"))
)

# Útil si se quiere revisar si VEHICULOS o PERSONAS traen alguna variante extra
display(dist_por_tabla)


=== Distintos 'tipo_eve_norm' con conteo (HECHOS) ===
+-----------------+-----+
|tipo_eve_norm_str|count|
+-----------------+-----+
|COLISION         |34998|
|ATROPELLO        |21778|
|DERRAPE          |8826 |
|NULL             |6324 |
|CHOQUE           |2808 |
|VUELCO           |1689 |
|EMBARRANCO       |294  |
|CAIDA            |25   |
|IGNORADO         |15   |
|ENCUNET?         |2    |
+-----------------+-----+


=== Valores distintos (solo listado) ===
+-----------------+
|tipo_eve_norm_str|
+-----------------+
|NULL             |
|ATROPELLO        |
|CAIDA            |
|CHOQUE           |
|COLISION         |
|DERRAPE          |
|EMBARRANCO       |
|ENCUNET?         |
|IGNORADO         |
|VUELCO           |
+-----------------+


=== Aparición por año (tipo_eve_norm) ===
+---------+-----------------+-----+
|anio_norm|tipo_eve_norm_str|count|
+---------+-----------------+-----+
|2013     |NULL             |6324 |
|2014     |COLISION         |2345 |
|2014     |ATROPELLO        |2276 |

tipo_eve_norm_str count COLISION 34998 ATROPELLO 21778 DERRAPE 8826 null 6324 CHOQUE 2808 VUELCO 1689 EMBARRANCO 294 CAIDA 25 IGNORADO 15 ENCUNET? 2

anio_norm tipo_eve_norm_str count 2013 null 6324 2014 COLISION 2345 2014 ATROPELLO 2276 2014 DERRAPE 532 2014 CHOQUE 366 2014 VUELCO 132 2015 COLISION 3006 2015 ATROPELLO 2565 2015 DERRAPE 715 2015 CHOQUE 438 2015 VUELCO 130 2016 COLISION 3595 2016 ATROPELLO 2766 2016 DERRAPE 820 2016 VUELCO 532 2016 CHOQUE 243 2016 IGNORADO 8 2017 COLISION 2734 2017 ATROPELLO 2077 2017 DERRAPE 646 2017 VUELCO 259 2017 CHOQUE 163 2018 COLISION 3059 2018 ATROPELLO 2060 2018 DERRAPE 848 2018 CHOQUE 197 2018 VUELCO 168 2018 EMBARRANCO 59 2018 CAIDA 4 2019 COLISION 3623 2019 ATROPELLO 2102 2019 DERRAPE 1000 2019 CHOQUE 188 2019 VUELCO 131 2019 IGNORADO 3 2020 COLISION 3614 2020 ATROPELLO 1608 2020 DERRAPE 909 2020 CHOQUE 140 2020 EMBARRANCO 45 2020 VUELCO 28 2020 CAIDA 6 2021 COLISION 4473 2021 ATROPELLO 2064 2021 DERRAPE 1374 2021 CHOQUE 158 2021 EMBARRANCO 68 2021 CAIDA 12 2021 IGNORADO 4 2022 COLISION 4230 2022 ATROPELLO 2082 2022 DERRAPE 907 2022 CHOQUE 472 2022 VUELCO 176 2022 EMBARRANCO 52 2022 CAIDA 3 2022 ENCUNET? 2 2023 COLISION 4319 2023 ATROPELLO 2178 2023 DERRAPE 1075 2023 CHOQUE 443 2023 VUELCO 133 2023 EMBARRANCO 70

source_table tipo_eve_norm_str count HECHOS COLISION 34998 HECHOS ATROPELLO 21778 HECHOS DERRAPE 8826 HECHOS null 6324 HECHOS CHOQUE 2808 HECHOS VUELCO 1689 HECHOS EMBARRANCO 294 HECHOS CAIDA 25 HECHOS IGNORADO 15 HECHOS ENCUNET? 2 PERSONAS COLISION 49695 PERSONAS ATROPELLO 24638 PERSONAS DERRAPE 12327 PERSONAS null 9060 PERSONAS CHOQUE 8460 PERSONAS VUELCO 4074 PERSONAS EMBARRANCO 1105 PERSONAS CAIDA 30 PERSONAS IGNORADO 28 VEHICULOS COLISION 69806 VEHICULOS ATROPELLO 21787 VEHICULOS DERRAPE 9139 VEHICULOS null 6323 VEHICULOS CHOQUE 2881 VEHICULOS VUELCO 1706 VEHICULOS EMBARRANCO 297 VEHICULOS CAIDA 25 VEHICULOS IGNORADO 21 VEHICULOS ENCUNETO 3

# Inciso 3 — Tipos de accidente (master_unico)

## Resumen ejecutivo
- En **HECHOS** predominan tres categorías: **COLISION** (45.6%), **ATROPELLO** (28.4%) y **DERRAPE** (11.5%).  
- **2013** aparece con **tipo no clasificado** (NULL) en HECHOS (6,324 filas), lo que explica la cuarta “categoría” por volumen.
- Categorías de baja frecuencia: **CHOQUE**, **VUELCO**, **EMBARRANCO**, **CAIDA**, **ENCUNET?** e **IGNORADO**.
- Se observan **variantes ortográficas/acentos** y **códigos residuales** (p. ej., `ENCUNET?` vs `ENCUNETO`), que conviene normalizar.

---

## Distribución general (HECHOS)
**Top de tipos (con conteos):**
- **COLISION:** 34,998  
- **ATROPELLO:** 21,778  
- **DERRAPE:** 8,826  
- **NULL (no clasificado):** 6,324  
- **CHOQUE:** 2,808  
- **VUELCO:** 1,689  
- **EMBARRANCO:** 294  
- **CAIDA:** 25  
- **IGNORADO:** 15  
- **ENCUNET?**: 2  

> Suma total HECHOS = 76,759 (coincide con el total de filas reportado).

**Valores distintos detectados:**
`{COLISION, ATROPELLO, DERRAPE, CHOQUE, VUELCO, EMBARRANCO, CAIDA, ENCUNET?, IGNORADO, NULL}`

---

## Evolución anual (HECHOS)
- **2013:** todas las filas aparecen como **NULL** en `tipo_eve_norm` (6,324). Indica ausencia de clasificación en la fuente original de ese año.
- **2014–2019:** **COLISION** y **ATROPELLO** se mantienen como los principales tipos; **DERRAPE** permanece tercero.  
- **2020 (pandemia):** **ATROPELLO** baja a 1,608; **COLISION** se mantiene (3,614). Es consistente con menor movilidad peatonal y cambios de patrón de tránsito.
- **2021:** pico reciente de **COLISION** (4,473) y repunte general tras 2020.  
- **2022–2023:** la jerarquía se conserva (**COLISION > ATROPELLO > DERRAPE**) con ligeras variaciones.  
- **Casos residuales por año:**  
  - **IGNORADO:** solo aparece en 2016 (8), 2019 (3) y 2021 (4).  
  - **EMBARRANCO/CAIDA/ENCUNET?**: muy esporádicos, con recuentos bajos (p. ej., **ENCUNET?** solo 2 en 2022).

---

## Consistencia entre tablas del master
- En **PERSONAS** y **VEHICULOS** también domina **COLISION**, seguido de **ATROPELLO** y **DERRAPE**.  
- Los **volúmenes mayores** en PERSONAS/VEHICULOS respecto a HECHOS son esperables (múltiples personas/vehículos por accidente).  
- Se detecta **inconsistencia de rótulos**:
  - HECHOS: `ENCUNET?`  
  - VEHICULOS: `ENCUNETO`  
  → Ambas deben mapearse a un **único valor estándar** (sugerencia: **ENCUNETADO** o **ENCUNETAMIENTO**).

---

## Calidad de datos y normalización recomendada
1. **Acentos/ortografía**  
   - Unificar **COLISION ↔ COLISIÓN**, **EMBARRANCO ↔ EMBARRANCÓ**, **CAIDA ↔ CAÍDA**, etc., en **mayúsculas sin acento** para estabilidad (o, si se prefiere, con acento pero de forma **consistente**).
2. **Valores residuales**  
   - Mapear `ENCUNET?`/`ENCUNETO` → **ENCUNETADO** (o el estándar acordado).  
   - Mantener **IGNORADO** como categoría explícita o reclasificarla en **OTRO/IGNORADO** si se desea limpiar análisis.
3. **Valores nulos**  
   - **2013:** `tipo_eve_norm` es **NULL** en HECHOS. Para análisis por tipo, **excluir** o **separar** 2013 como “no clasificado”.  
   - Documentar este sesgo cuando se comparen proporciones por tipo a través de años.

---

## Implicaciones analíticas
- Para métricas comparables en el tiempo (incisos 5, 8, 9, 12, etc.), conviene:
  - **Filtrar `tipo_eve_norm` no nulo** o **imputar** 2013 como “DESCONOCIDO/NO CLASIFICADO” (no mezclar con tipos reales).  
  - **Aplicar el diccionario de normalización** antes de agrupar (asegura que no se “divida” una misma categoría por diferencias de escritura).
- La tríada **COLISION–ATROPELLO–DERRAPE** concentra >85% de HECHOS (sin 2013 nulos), por lo que explica la mayor parte de las tendencias.




## **Inciso 4**

In [0]:
# INCISO 4 — ¿Cuántos departamentos únicos aparecen en las bases?
# Usa la tabla unificada "master_unico" ya cargada en la sesión.

from pyspark.sql import functions as F

# 1) DataFrame base
df = spark.table("master_unico")

# Normalización suave del nombre (por si hay espacios/acentos inconsistentes)
df_norm = (
    df.withColumn("depto_name_std", F.upper(F.trim(F.col("depto_name_std"))))
)

# 2) Conteo por 'source_table' (HECHOS, PERSONAS, VEHICULOS)
by_source = (
    df_norm.groupBy("source_table")
           .agg(
               F.countDistinct("depto_code").alias("distinct_depto_code"),
               F.countDistinct("depto_name_std").alias("distinct_depto_name_std")
           )
           .orderBy("source_table")
)

print("=== Departamentos únicos por base (códigos y nombres) ===")
display(by_source)

# 3) Conteo TOTAL en el master_unico (sin separar por base)
total_dist = df_norm.agg(
    F.countDistinct("depto_code").alias("distinct_depto_code_total"),
    F.countDistinct("depto_name_std").alias("distinct_depto_name_std_total")
)

print("\n=== Departamentos únicos en TODAS las bases (total master_unico) ===")
total_dist.show(truncate=False)

# 4) (Opcional) Listado detallado para revisión de cobertura
print("\n=== Listado de CÓDIGOS de departamento (con frecuencia) ===")
codes_counts = (
    df_norm.filter(F.col("depto_code").isNotNull())
           .groupBy("depto_code")
           .agg(F.count("*").alias("count"))
           .orderBy(F.col("depto_code").cast("int"))
)
display(codes_counts)

print("\n=== Listado de NOMBRES normalizados (con frecuencia) ===")
names_counts = (
    df_norm.filter(F.col("depto_name_std").isNotNull())
           .groupBy("depto_name_std")
           .agg(F.count("*").alias("count"))
           .orderBy("depto_name_std")
)
display(names_counts)


=== Departamentos únicos por base (códigos y nombres) ===


source_table distinct_depto_code distinct_depto_name_std HECHOS 22 22 PERSONAS 22 22 VEHICULOS 22 22


=== Departamentos únicos en TODAS las bases (total master_unico) ===
+-------------------------+-----------------------------+
|distinct_depto_code_total|distinct_depto_name_std_total|
+-------------------------+-----------------------------+
|22                       |22                           |
+-------------------------+-----------------------------+


=== Listado de CÓDIGOS de departamento (con frecuencia) ===


depto_code count 1 85475 2 5679 3 7621 4 9361 5 18038 6 8202 7 4535 8 2307 9 8360 10 8231 11 6485 12 5974 13 5300 14 5691 15 4388 16 10718 17 8110 18 8530 19 6854 20 5564 21 3561 22 6682


=== Listado de NOMBRES normalizados (con frecuencia) ===


depto_name_std count ALTA VERAPAZ 10718 BAJA VERAPAZ 4388 CHIMALTENANGO 9361 CHIQUIMULA 5564 EL PROGRESO 5679 ESCUINTLA 18038 GUATEMALA 85475 HUEHUETENANGO 5300 IZABAL 8530 JALAPA 3561 JUTIAPA 6682 PETEN 8110 QUETZALTENANGO 8360 QUICHE 5691 RETALHULEU 6485 SACATEPEQUEZ 7621 SAN MARCOS 5974 SANTA ROSA 8202 SOLOLA 4535 SUCHITEPEQUEZ 8231 TOTONICAPAN 2307 ZACAPA 6854

## Inciso 4 — Departamentos únicos y distribución (master_unico)

### Cobertura por tabla
Los datos muestran **cobertura completa** de departamentos en las tres fuentes:
- **HECHOS:** 22 códigos únicos / 22 nombres únicos  
- **PERSONAS:** 22 / 22  
- **VEHÍCULOS:** 22 / 22  

En el total del `master_unico` también se mantiene la consistencia:
- **Total:** 22 códigos únicos / 22 nombres únicos

> Interpretación: las tres capas comparten el mismo universo de 22 departamentos; no hay departamentos “fantasma” exclusivos de alguna tabla ni faltantes en otra.

---

### Distribución por departamento (acumulado de todas las tablas)
**Top y bottom por volumen (cuentas sobre registros con `depto_code` no nulo):**

- **Más alto:**  
  - **GUATEMALA:** 85,475 (≈36% del total con código)  
  - **ESCUINTLA:** 18,038  
  - **ALTA VERAPAZ:** 10,718  
  - **IZABAL:** 8,530  
  - **PETEN:** 8,110  
  - **QUETZALTENANGO:** 8,360  
  - **SUCHITEPEQUEZ:** 8,231  

- **Más bajo:**  
  - **TOTONICAPAN:** 2,307  
  - **JALAPA:** 3,561  
  - **SOLOLA:** 4,535  
  - **BAJA VERAPAZ:** 4,388  

> Observación principal: existe un **sesgo metropolitano** evidente; el departamento de Guatemala concentra ~**36%** de todos los registros que traen código de departamento. Esto es consistente con mayor densidad poblacional, parque vehicular y flujo vial.

---

### Consistencia código–nombre
- Los **nombres están normalizados** (mayúsculas y sin acentos): `PETEN`, `QUICHE`, `SACATEPEQUEZ`, etc.  
- **No aparecen variantes duplicadas** (p. ej., `PETÉN` vs `PETEN`), lo que **evita doble conteo** en agregaciones por nombre.
- El conteo agregado por `depto_code` (235,666) coincide con la suma de filas con código presente en cada subtabla:
  - HECHOS (`depto_code` no nulo): 60,617  
  - VEHÍCULOS: 87,552  
  - PERSONAS: 87,497  
  - **Total:** 235,666

> Implicación: se puede agrupar indistintamente por `depto_code` o por `depto_name_std` sin efectos de duplicidad por etiquetas.

---

### Recomendaciones para los incisos siguientes
- Usar **`depto_code`** como clave primaria territorial en joins y agregaciones, y **mostrar `depto_name_std`** solo como etiqueta.
- Mantener los filtros de **no nulos** en `depto_code` para que las tasas y comparativos entre departamentos sean consistentes.
- Para visualizaciones, ordenar por conteo descendente para resaltar la asimetría (Guatemala vs resto).

---
**Tabla de apoyo (cuentas por `depto_code`):**
- 1: 85,475  
- 2: 5,679  
- 3: 7,621  
- 4: 9,361  
- 5: 18,038  
- 6: 8,202  
- 7: 4,535  
- 8: 2,307  
- 9: 8,360  
- 10: 8,231  
- 11: 6,485  
- 12: 5,974  
- 13: 5,300  
- 14: 5,691  
- 15: 4,388  
- 16: 10,718  
- 17: 8,110  
- 18: 8,530  
- 19: 6,854  
- 20: 5,564  
- 21: 3,561  
- 22: 6,682

**Tabla de apoyo (cuentas por `depto_name_std`):**
- GUATEMALA: 85,475  
- ESCUINTLA: 18,038  
- ALTA VERAPAZ: 10,718  
- IZABAL: 8,530  
- PETEN: 8,110  
- QUETZALTENANGO: 8,360  
- SUCHITEPEQUEZ: 8,231  
- SAN MARCOS: 5,974  
- RETALHULEU: 6,485  
- QUICHE: 5,691  
- HUEHUETENANGO: 5,300  
- CHIMALTENANGO: 9,361  
- SANTA ROSA: 8,202  
- JUTIAPA: 6,682  
- ZACAPA: 6,854  
- CHIQUIMULA: 5,564  
- SACATEPEQUEZ: 7,621  
- BAJA VERAPAZ: 4,388  
- SOLOLA: 4,535  
- JALAPA: 3,561  
- SUCHITEPEQUEZ: 8,231  
- TOTONICAPAN: 2,307



## **Inciso 5**

In [0]:
# INCISO 5 — Total de ACCIDENTES por AÑO y DEPARTAMENTO (usando la tabla HECHOS)

from pyspark.sql import functions as F

# 1) Agregar: accidentes por año y departamento
accidentes_anio_depto = (
    spark.table("hechos")  # vista creada en el inciso 1
    .select("anio_norm", "depto_code", "depto_name_std")
    .where(F.col("anio_norm").isNotNull() & F.col("depto_code").isNotNull())
    .groupBy("anio_norm", "depto_code", "depto_name_std")
    .agg(F.count(F.lit(1)).alias("accidentes"))
)

# 2) Mostrar resultado y usar 'display' como gráfico de barras
#    En la UI: cambiar a "Plot" -> "Bar chart"
#      - Keys (x): anio_norm
#      - Values (y): accidentes
#      - Series group / Group by: depto_name_std (o depto_code)
display(accidentes_anio_depto.orderBy("anio_norm", "depto_code"))

# (Opcional) Versión más legible: solo TOP 10 departamentos por volumen total
top10_deptos = (
    accidentes_anio_depto.groupBy("depto_code", "depto_name_std")
    .agg(F.sum("accidentes").alias("total_depto"))
    .orderBy(F.desc("total_depto"))
    .limit(10)
)

accidentes_top10 = accidentes_anio_depto.join(F.broadcast(top10_deptos.select("depto_code")), on="depto_code")
display(accidentes_top10.orderBy("anio_norm", F.desc("accidentes")))

# (Sanity check) El total por año debe coincidir con el conteo de HECHOS por año:
totales_por_anio = accidentes_anio_depto.groupBy("anio_norm").agg(F.sum("accidentes").alias("accidentes"))
display(totales_por_anio.orderBy("anio_norm"))


anio_norm depto_code depto_name_std accidentes 2013 1 GUATEMALA 2267 2013 10 SUCHITEPEQUEZ 269 2013 11 RETALHULEU 133 2013 12 SAN MARCOS 221 2013 13 HUEHUETENANGO 201 2013 14 QUICHE 151 2013 15 BAJA VERAPAZ 100 2013 16 ALTA VERAPAZ 219 2013 17 PETEN 214 2013 18 IZABAL 234 2013 19 ZACAPA 229 2013 2 EL PROGRESO 132 2013 20 CHIQUIMULA 204 2013 21 JALAPA 89 2013 22 JUTIAPA 174 2013 3 SACATEPEQUEZ 142 2013 4 CHIMALTENANGO 212 2013 5 ESCUINTLA 424 2013 6 SANTA ROSA 214 2013 7 SOLOLA 126 2013 8 TOTONICAPAN 81 2013 9 QUETZALTENANGO 288 2014 1 GUATEMALA 1912 2014 10 SUCHITEPEQUEZ 223 2014 11 RETALHULEU 132 2014 12 SAN MARCOS 154 2014 13 HUEHUETENANGO 136 2014 14 QUICHE 144 2014 15 BAJA VERAPAZ 95 2014 16 ALTA VERAPAZ 223 2014 17 PETEN 219 2014 18 IZABAL 255 2014 19 ZACAPA 148 2014 2 EL PROGRESO 127 2014 20 CHIQUIMULA 126 2014 21 JALAPA 75 2014 22 JUTIAPA 172 2014 3 SACATEPEQUEZ 150 2014 4 CHIMALTENANGO 228 2014 5 ESCUINTLA 444 2014 6 SANTA ROSA 242 2014 7 SOLOLA 109 2014 8 TOTONICAPAN 82 2014 9 QUETZALTENANGO 255 2015 1 GUATEMALA 2273 2015 10 SUCHITEPEQUEZ 243 2015 11 RETALHULEU 181 2015 12 SAN MARCOS 177 2015 13 HUEHUETENANGO 173 2015 14 QUICHE 157 2015 15 BAJA VERAPAZ 112 2015 16 ALTA VERAPAZ 315 2015 17 PETEN 288 2015 18 IZABAL 285 2015 19 ZACAPA 208 2015 2 EL PROGRESO 160 2015 20 CHIQUIMULA 172 2015 21 JALAPA 96 2015 22 JUTIAPA 182 2015 3 SACATEPEQUEZ 209 2015 4 CHIMALTENANGO 263 2015 5 ESCUINTLA 569 2015 6 SANTA ROSA 254 2015 7 SOLOLA 152 2015 8 TOTONICAPAN 69 2015 9 QUETZALTENANGO 316 2016 1 GUATEMALA 2619 2016 10 SUCHITEPEQUEZ 259 2016 11 RETALHULEU 262 2016 12 SAN MARCOS 198 2016 13 HUEHUETENANGO 225 2016 14 QUICHE 195 2016 15 BAJA VERAPAZ 149 2016 16 ALTA VERAPAZ 353 2016 17 PETEN 267 2016 18 IZABAL 286 2016 19 ZACAPA 315 2016 2 EL PROGRESO 201 2016 20 CHIQUIMULA 181 2016 21 JALAPA 98 2016 22 JUTIAPA 180 2016 3 SACATEPEQUEZ 287 2016 4 CHIMALTENANGO 420 2016 5 ESCUINTLA 627 2016 6 SANTA ROSA 261 2016 7 SOLOLA 182 2016 8 TOTONICAPAN 76 2016 9 QUETZALTENANGO 323 2017 1 GUATEMALA 2128 2017 10 SUCHITEPEQUEZ 238 2017 11 RETALHULEU 148 2017 12 SAN MARCOS 130 2017 13 HUEHUETENANGO 140 2017 14 QUICHE 145 2017 15 BAJA VERAPAZ 111 2017 16 ALTA VERAPAZ 247 2017 17 PETEN 194 2017 18 IZABAL 200 2017 19 ZACAPA 184 2017 2 EL PROGRESO 139 2017 20 CHIQUIMULA 114 2017 21 JALAPA 92 2017 22 JUTIAPA 177 2017 3 SACATEPEQUEZ 213 2017 4 CHIMALTENANGO 244 2017 5 ESCUINTLA 461 2017 6 SANTA ROSA 234 2017 7 SOLOLA 124 2017 8 TOTONICAPAN 44 2017 9 QUETZALTENANGO 172 2018 1 GUATEMALA 2722 2018 10 SUCHITEPEQUEZ 205 2018 11 RETALHULEU 153 2018 12 SAN MARCOS 147 2018 13 HUEHUETENANGO 108 2018 14 QUICHE 139 2018 15 BAJA VERAPAZ 132 2018 16 ALTA VERAPAZ 265 2018 17 PETEN 187 2018 18 IZABAL 210 2018 19 ZACAPA 184 2018 2 EL PROGRESO 168 2018 20 CHIQUIMULA 193 2018 21 JALAPA 78 2018 22 JUTIAPA 160 2018 3 SACATEPEQUEZ 201 2018 4 CHIMALTENANGO 238 2018 5 ESCUINTLA 407 2018 6 SANTA ROSA 192 2018 7 SOLOLA 104 2018 8 TOTONICAPAN 47 2018 9 QUETZALTENANGO 155 2019 1 GUATEMALA 3017 2019 10 SUCHITEPEQUEZ 218 2019 11 RETALHULEU 207 2019 12 SAN MARCOS 149 2019 13 HUEHUETENANGO 99 2019 14 QUICHE 114 2019 15 BAJA VERAPAZ 129 2019 16 ALTA VERAPAZ 317 2019 17 PETEN 217 2019 18 IZABAL 234 2019 19 ZACAPA 156 2019 2 EL PROGRESO 151 2019 20 CHIQUIMULA 153 2019 21 JALAPA 94 2019 22 JUTIAPA 219 2019 3 SACATEPEQUEZ 233 2019 4 CHIMALTENANGO 274 2019 5 ESCUINTLA 452 2019 6 SANTA ROSA 220 2019 7 SOLOLA 140 2019 8 TOTONICAPAN 63 2019 9 QUETZALTENANGO 191 2020 1 GUATEMALA 2676 2020 10 SUCHITEPEQUEZ 201 2020 11 RETALHULEU 203 2020 12 SAN MARCOS 147 2020 13 HUEHUETENANGO 113 2020 14 QUICHE 106 2020 15 BAJA VERAPAZ 96 2020 16 ALTA VERAPAZ 296 2020 17 PETEN 206 2020 18 IZABAL 230 2020 19 ZACAPA 141 2020 2 EL PROGRESO 124 2020 20 CHIQUIMULA 130 2020 21 JALAPA 121 2020 22 JUTIAPA 165 2020 3 SACATEPEQUEZ 202 2020 4 CHIMALTENANGO 229 2020 5 ESCUINTLA 440 2020 6 SANTA ROSA 197 2020 7 SOLOLA 81 2020 8 TOTONICAPAN 49 2020 9 QUETZALTENANGO 197 2021 1 GUATEMALA 3450 2021 10 SUCHITEPEQUEZ 229 2021 11 RETA

depto_code anio_norm depto_name_std accidentes 1 2013 GUATEMALA 2267 5 2013 ESCUINTLA 424 9 2013 QUETZALTENANGO 288 10 2013 SUCHITEPEQUEZ 269 18 2013 IZABAL 234 16 2013 ALTA VERAPAZ 219 17 2013 PETEN 214 6 2013 SANTA ROSA 214 4 2013 CHIMALTENANGO 212 3 2013 SACATEPEQUEZ 142 1 2014 GUATEMALA 1912 5 2014 ESCUINTLA 444 9 2014 QUETZALTENANGO 255 18 2014 IZABAL 255 6 2014 SANTA ROSA 242 4 2014 CHIMALTENANGO 228 16 2014 ALTA VERAPAZ 223 10 2014 SUCHITEPEQUEZ 223 17 2014 PETEN 219 3 2014 SACATEPEQUEZ 150 1 2015 GUATEMALA 2273 5 2015 ESCUINTLA 569 9 2015 QUETZALTENANGO 316 16 2015 ALTA VERAPAZ 315 17 2015 PETEN 288 18 2015 IZABAL 285 4 2015 CHIMALTENANGO 263 6 2015 SANTA ROSA 254 10 2015 SUCHITEPEQUEZ 243 3 2015 SACATEPEQUEZ 209 1 2016 GUATEMALA 2619 5 2016 ESCUINTLA 627 4 2016 CHIMALTENANGO 420 16 2016 ALTA VERAPAZ 353 9 2016 QUETZALTENANGO 323 3 2016 SACATEPEQUEZ 287 18 2016 IZABAL 286 17 2016 PETEN 267 6 2016 SANTA ROSA 261 10 2016 SUCHITEPEQUEZ 259 1 2017 GUATEMALA 2128 5 2017 ESCUINTLA 461 16 2017 ALTA VERAPAZ 247 4 2017 CHIMALTENANGO 244 10 2017 SUCHITEPEQUEZ 238 6 2017 SANTA ROSA 234 3 2017 SACATEPEQUEZ 213 18 2017 IZABAL 200 17 2017 PETEN 194 9 2017 QUETZALTENANGO 172 1 2018 GUATEMALA 2722 5 2018 ESCUINTLA 407 16 2018 ALTA VERAPAZ 265 4 2018 CHIMALTENANGO 238 18 2018 IZABAL 210 10 2018 SUCHITEPEQUEZ 205 3 2018 SACATEPEQUEZ 201 6 2018 SANTA ROSA 192 17 2018 PETEN 187 9 2018 QUETZALTENANGO 155 1 2019 GUATEMALA 3017 5 2019 ESCUINTLA 452 16 2019 ALTA VERAPAZ 317 4 2019 CHIMALTENANGO 274 18 2019 IZABAL 234 3 2019 SACATEPEQUEZ 233 6 2019 SANTA ROSA 220 10 2019 SUCHITEPEQUEZ 218 17 2019 PETEN 217 9 2019 QUETZALTENANGO 191 1 2020 GUATEMALA 2676 5 2020 ESCUINTLA 440 16 2020 ALTA VERAPAZ 296 18 2020 IZABAL 230 4 2020 CHIMALTENANGO 229 17 2020 PETEN 206 3 2020 SACATEPEQUEZ 202 10 2020 SUCHITEPEQUEZ 201 9 2020 QUETZALTENANGO 197 6 2020 SANTA ROSA 197 1 2021 GUATEMALA 3450 5 2021 ESCUINTLA 615 16 2021 ALTA VERAPAZ 380 4 2021 CHIMALTENANGO 297 3 2021 SACATEPEQUEZ 270 17 2021 PETEN 266 9 2021 QUETZALTENANGO 253 6 2021 SANTA ROSA 238 18 2021 IZABAL 237 10 2021 SUCHITEPEQUEZ 229

anio_norm accidentes 2013 6324 2014 5651 2015 6854 2016 7964 2017 5879 2018 6395 2019 7047 2020 6350 2021 8153

# Inciso 5 — Análisis de accidentes por **año** y **departamento**

## Resumen de volumen anual (HECHOS)
Los conteos anuales reportados muestran una evolución con caída en 2020 y pico en 2021:

- **2013:** 6,324  
- **2014:** 5,651  
- **2015:** 6,854  
- **2016:** 7,964  
- **2017:** 5,879  
- **2018:** 6,395  
- **2019:** 7,047  
- **2020:** 6,350  
- **2021:** 8,153  ← mayor valor del periodo

**Lectura rápida:** el descenso de 2020 es consistente con restricciones de movilidad; 2021 rebota a un máximo.

---

## Patrones por departamento (2013–2021)

### Concentración metropolitana
- **Guatemala (Depto 1)** domina todos los años:  
  2013: 2,267 → 2014: 1,912 → 2015: 2,273 → 2016: 2,619 → 2017: 2,128 →  
  2018: 2,722 → 2019: 3,017 → 2020: 2,676 → **2021: 3,450**.  
  **Interpretación:** mayor densidad vehicular y flujo urbano sostienen el liderazgo y el rebote post-2020.

### Segundo bloque de alta incidencia
- **Escuintla (5)** se mantiene alto: 424 → 444 → 569 → **627** → 461 → 407 → 452 → 440 → **615**.  
  Presenta un **pico en 2016** y un **nuevo máximo en 2021**.
- **Quetzaltenango (9)** y **Suchitepéquez (10)** conservan volúmenes medios-altos y estables (p. ej., Quetzaltenango: 288, 255, 316, 323, 172, 155, 191, 197, 253).

### Departamentos con tendencia al alza moderada
- **Alta Verapaz (16)**: 219 → 223 → 315 → 353 → 247 → 265 → 317 → 296 → **380**.  
  **Lectura:** incremento estructural tras 2015–2016 y nuevo máximo en 2021.

### Cola baja y estabilidad
- **Totonicapán (8)** es el de menor incidencia en casi todos los años:  
  81 → 82 → **69** → 76 → **44** → 47 → 63 → 49 → 77.  
- **Jalapa (21)** y **Sololá (7)** también se ubican sistemáticamente en la parte baja del ranking.

---

## Picos por año (muestras)
- **2013:** Guatemala (2,267) muy por encima; Escuintla (424) y Quetzaltenango (288) siguen.  
- **2016:** máximos anuales en varios departamentos (p. ej., Escuintla 627, Chimaltenango 420).  
- **2021:** máximos recientes generalizados (Guatemala 3,450; Escuintla 615; Alta Verapaz 380).

---

## Señales temporales relevantes
1. **Efecto 2020:** caída transversal (p. ej., Guatemala 3,017 → 2,676) coherente con menor movilidad.  
2. **Rebote 2021:** incremento notable en casi todos los departamentos, alcanzando máximos del periodo.  
3. **Persistencia espacial:** la jerarquía territorial cambia poco; Guatemala y Escuintla sostienen el podio.

---

## Posibles factores explicativos
- **Exposición al riesgo:** densidad poblacional, parque vehicular y ejes viales nacionales (Guatemala, Escuintla).  
- **Actividad económica y logística:** corredores de carga y puertos incrementan flujo (Escuintla, Izabal).  
- **Orografía y clima:** tramos con pendientes/curvas o lluvias pueden elevar siniestralidad (Alta Verapaz).  
- **Políticas y controles:** variaciones locales en fiscalización o infraestructura inciden en tendencias.

---

## Recomendaciones de visualización (en Databricks `display`)
- **Barras apiladas:** `anio_norm` en el eje X, `accidentes` en Y y **serie** por `depto_name_std` para comparar composición anual.  
- **Barras agrupadas:** filtrar a los **Top 10** departamentos por volumen total para claridad.  
- **Línea anual:** sumar por año y graficar tendencia global (confirma el valle 2020 y el pico 2021).

---

## Notas de calidad
- Los totales por año reflejados concuerdan con la suma por departamentos del mismo año en las salidas mostradas.  
- La cobertura del análisis se centra en 2013–2021 según el extracto entregado; si existen 2022–2023, conviene incluirlos en la misma estructura para completar la serie.


## **Inciso 6**

In [0]:
# INCISO 6 (corregido) — Día de la semana con más accidentes.
# Recalcula el día de la semana desde (anio_norm, mes_norm, dia_norm).
from pyspark.sql import functions as F

hechos = spark.table("hechos")

# 1) Resolver año objetivo (2024 si existe; si no, último disponible)
years = (
    hechos
    .select(F.col("anio_norm").cast("int").alias("anio"))
    .where(F.col("anio").isNotNull())
    .distinct()
    .orderBy("anio")
    .collect()
)
years_list = [r["anio"] for r in years]
if not years_list:
    raise ValueError("No hay años disponibles en 'hechos'.")
target_year = 2024 if 2024 in years_list else max(years_list)

# 2) Construir fecha válida y filtrar al año objetivo
base = (
    hechos
    .withColumn("anio_i", F.col("anio_norm").cast("int"))
    .withColumn("mes_i",  F.col("mes_norm").cast("int"))
    .withColumn("dia_i",  F.col("dia_norm").cast("int"))
    .withColumn("fecha",  F.to_date(F.concat_ws("-", "anio_i", "mes_i", "dia_i"), "yyyy-M-d"))
    .where(F.col("fecha").isNotNull())
    .where(F.year("fecha") == F.lit(target_year))
)

# 3) dayofweek(fecha): Spark da 1=Domingo..7=Sábado.
#    Convertimos a lunes=1..domingo=7 para mantener el orden habitual.
dow_sun1 = F.dayofweek("fecha")
dow_mon1 = ((dow_sun1 + F.lit(5)) % F.lit(7)) + F.lit(1)

dist_dow = (
    base
    .withColumn("dow_num", dow_mon1.cast("int"))
    .withColumn(
        "dow_label",
        F.when(F.col("dow_num")==1,"LUNES")
         .when(F.col("dow_num")==2,"MARTES")
         .when(F.col("dow_num")==3,"MIÉRCOLES")
         .when(F.col("dow_num")==4,"JUEVES")
         .when(F.col("dow_num")==5,"VIERNES")
         .when(F.col("dow_num")==6,"SÁBADO")
         .when(F.col("dow_num")==7,"DOMINGO")
    )
    .groupBy("dow_num", "dow_label")
    .agg(F.count(F.lit(1)).alias("accidentes"))
    .orderBy("dow_num")
)

print(f"[INFO] Año analizado: {target_year}")
display(dist_dow.select("dow_num", "dow_label", "accidentes"))  # Configura como 'Bar chart': X=dow_label, Y=accidentes

# 4) Día con más accidentes
top = dist_dow.orderBy(F.desc("accidentes")).limit(1).collect()
if top:
    print(f"Día con más accidentes en {target_year}: {top[0]['dow_label']} ({top[0]['accidentes']})")
else:
    print(f"No hay datos para {target_year} tras construir la fecha.")


[INFO] Año analizado: 2023


dow_num dow_label accidentes 1 LUNES 1053 2 MARTES 932 3 MIÉRCOLES 924 4 JUEVES 959 5 VIERNES 1173 6 SÁBADO 1557 7 DOMINGO 1620

Día con más accidentes en 2023: DOMINGO (1620)


# Inciso 6 — Accidentes por día de la semana (análisis)

## Resumen
- **Total analizado:** 8,218 accidentes.
- **Día con más accidentes:** **DOMINGO (1,620; 19.7%)**.
- **Segundo lugar:** **SÁBADO (1,557; 18.9%)**.
- **Tercer lugar:** **VIERNES (1,173; 14.3%)**.
- **Menor registro:** **MIÉRCOLES (924; 11.2%)**.

## Tabla de distribución
| Día         | Accidentes | % del total |
|-------------|------------:|------------:|
| LUNES       | 1,053       | 12.8%       |
| MARTES      |   932       | 11.3%       |
| MIÉRCOLES   |   924       | 11.2%       |
| JUEVES      |   959       | 11.7%       |
| VIERNES     | 1,173       | 14.3%       |
| SÁBADO      | 1,557       | 18.9%       |
| **DOMINGO** | **1,620**   | **19.7%**   |

## Patrones clave
- **Finde (Sábado+Domingo) = 3,177 accidentes (38.7%)** del total.
- **Viernes+Finde = 4,350 (52.9%)**, es decir, **más de la mitad** ocurre entre viernes y domingo.
- El **DOMINGO** supera en **≈67.5%** el promedio diario de lunes a jueves.
- La brecha **DOMINGO vs. día mínimo (MIÉRCOLES)** es de **+696 accidentes**.

## Posibles razones del pico en fin de semana
1. **Movilidad recreativa y social:** mayor volumen de viajes no laborales (salidas familiares, ocio, turismo de cercanía).
2. **Fatiga y conducción nocturna:** retornos del fin de semana, trayectos largos y conducción en la noche/tarde-noche.
3. **Consumo de alcohol y distracciones:** mayor exposición a factores de riesgo en actividades sociales.
4. **Cambios en la mezcla vehicular:** más motocicletas y autos particulares en ocio, distintos patrones a los días laborales.
5. **Flujos de retorno los domingos:** concentración de viajes de regreso hacia áreas urbanas genera **picos vespertinos/nocturnos**.

## Recomendaciones operativas
- **Refuerzo de controles** (alcoholímetro, velocidad) y **señalización** en **viernes noche a domingo noche**.
- **Campañas preventivas** enfocadas en **retorno dominical** (descanso, pausas, conductor designado).
- **Gestión de tráfico y patrullaje** focalizado en **corredores de salida/retorno** y zonas de ocio.




In [0]:
# INCISO 7 — Distribución de accidentes por hora del día en el Municipio de Guatemala
# Dataset: master_unico (con columnas normalizadas)
# Criterio de filtro: Municipio de Guatemala → depto_code = 1 y muni_code = 101

from pyspark.sql import functions as F

# 1) Filtrar hechos del Municipio de Guatemala y sanear hora
hechos = spark.table("master_unico").where(F.col("source_table") == "HECHOS")

hechos_gua = (
    hechos
    .where( (F.col("depto_code") == 1) & (F.col("muni_code") == 101) )
    .withColumn("hora_clean_int", F.col("hora_clean").cast("int"))
    .filter(F.col("hora_clean_int").between(0, 23))
)

print(f"[INFO] Municipio de Guatemala — filas consideradas: {hechos_gua.count():,}")

# 2) Distribución por hora (conteo)
dist_hora = (
    hechos_gua
    .groupBy("hora_clean_int")
    .agg(F.count(F.lit(1)).alias("accidentes"))
    .orderBy("hora_clean_int")
)

# 3) Mostrar tabla y (en Databricks) graficar
# En la tabla resultante, haga clic en 'Plot' → 'Bar chart'
#    - Keys (X): hora_clean_int
#    - Values (Y): accidentes
#    - Aggregation: SUM (por defecto)
display(dist_hora)

# 4) Opción alternativa: histograma real de una sola columna (Databricks)
#    - Haga clic en 'Plot' → 'Histogram'
#    - Values: hora_clean
display(
    hechos_gua.select(F.col("hora_clean_int").alias("hora_clean"))
)

# 5) (Opcional) Top 3 horas con más accidentes (impresión rápida)
top3 = dist_hora.orderBy(F.desc("accidentes")).limit(3).collect()
if top3:
    print("Top 3 horas (Municipio de Guatemala):")
    for r in top3:
        print(f"  - {int(r['hora_clean_int']):02d}:00 → {r['accidentes']} accidentes")


[INFO] Municipio de Guatemala — filas consideradas: 13,003


hora_clean_int accidentes 0 560 1 569 2 452 3 377 4 284 5 330 6 358 7 451 8 408 9 453 10 512 11 462 12 501 13 429 14 492 15 593 16 505 17 654 18 681 19 764 20 768 21 828 22 774 23 798

hora_clean 5 15 7 22 11 17 7 9 0 5 7 8 15 9 1 23 1 21 9 21 2 23 19 20 4 13 16 4 1 16 20 18 23 3 23 0 2 2 2 0 16 12 17 21 8 17 16 0 22 0 4 4 20 17 23 18 10 12 9 6 7 8 9 23 11 22 21 0 3 0 6 8 1 2 8 11 18 9 20 2 3 18 11 0 2 1 16 3 22 13 18 18 22 11 20 19 13 6 21 17 2 20 1 22 14 7 22 2 23 21 21 0 8 9 16 5 11 3 17 14 22 21 1 23 0 3 23 17 9 21 22 18 10 19 2 0 9 21 10 13 23 8 20 18 22 10 15 17 1 17 3 2 21 6 1 21 18 22 21 22 1 0 2 9 16 18 10 11 23 11 21 23 23 21 2 17 19 3 21 0 16 6 9 23 2 0 4 18 23 23 10 19 6 6 23 19 20 22 1 19 1 1 14 21 1 2 2 1 2 2 22 23 19 8 4 20 0 20 13 21 19 21 2 17 14 22 5 17 10 11 19 0 1 10 23 23 20 17 11 7 3 7 9 15 17 9 20 20 15 2 5 4 22 5 15 6 2 5 6 7 9 1 17 17 1 16 5 22 12 13 22 15 7 20 17 6 8 19 3 2 10 10 11 10 21 19 22 19 4 21 1 23 10 21 9 20 21 22 2 2 22 4 15 19 17 19 23 14 12 9 22 16 1 17 2 8 18 21 5 12 15 15 15 15 3 18 14 21 22 1 19 16 12 19 2 23 17 23 10 5 23 23 23 17 15 9 11 2 1 18 18 23 11 20 3 9 4 13 21 16 19 6 23 3 22 14 23 3 20 19 15 21 2 6 3 12 22 22 18 22 11 12 10 23 21 23 9 15 6 1 0 22 9 2 18 22 8 22 13 18 1 21 23 0 8 2 2 0 5 23 18 22 15 1 17 18 1 19 6 5 10 4 21 2 10 20 16 15 18 4 1 20 7 19 20 8 22 5 23 2 4 18 6 10 3 1 14 21 17 1 7 13 16 4 1 7 11 10 15 5 21 4 21 23 0 22 2 4 3 22 2 3 3 0 23 16 7 20 7 10 10 3 8 22 1 23 0 3 5 16 16 15 23 20 14 9 17 18 16 22 18 19 8 20 10 14 23 6 23 21 13 21 20 18 2 2 5 10 18 21 11 21 6 15 21 23 9 15 16 2 23 19 0 7 21 10 23 9 18 18 7 22 11 14 13 13 15 10 20 3 1 20 10 3 3 6 0 10 22 19 18 23 9 6 2 6 13 4 1 23 20 23 17 0 19 20 11 15 17 12 11 14 15 22 13 23 3 22 19 3 1 20 15 21 18 12 2 16 17 20 21 16 16 2 9 23 1 2 1 3 5 1 2 9 22 22 1 11 15 21 23 23 14 11 16 23 8 17 12 16 1 15 1 18 19 23 6 12 12 2 20 12 20 23 6 19 21 21 21 4 17 17 20 21 23 11 16 8 0 6 2 20 17 15 21 11 21 6 23 23 16 16 20 15 12 19 12 8 23 21 14 20 23 10 21 2 4 5 0 16 21 3 18 16 21 1 19 18 3 4 1 21 18 11 5 3 17 11 0 14 18 11 23 2 5 1 20 21 10 18 0 4 5 18 3 5 14 19 16 21 23 17 19 7 15 21 21 1 20 22 0 1 17 1 8 20 13 1 0 14 11 11 20 23 1 5 5 19 2 15 23 0 9 11 21 21 20 11 3 3 15 2 4 8 2 0 21 10 10 15 1 0 22 13 22 5 0 14 15 2 19 19 15 22 21 0 2 19 8 23 13 8 3 5 11 22 10 22 16 20 7 10 0 7 11 19 22 21 23 11 18 1 0 22 9 4 14 23 1 21 21 0 3 15 9 2 0 21 7 9 13 5 23 3 15 5 1 4 18 3 22 2 3 6 6 16 21 3 23 17 19 21 20 21 18 21 19 17 22 18 5 9 18 23 21 16 19 14 20 11 13 0 2 8 20 0 17 23 21 13 12 15 0 0 17 13 7 10 7 20 16 19 13 13 12 15 16 20 17 0 10 10 20 17 16 15 12 8 15 20 9 18 23 6 19 13 14 20 7 14 8 6 22 19 2 21 17 17 18 17 1 22 22 21 15 14 19 20 20 18 23 23 19 13 9 12 20 20 8 22 10 6 6 10 22 18 20 21 18 8 18 9 8 8 7 23 23 23 18 18 16 19 19 15 15 21 21 19 14 0 23 16 19 18 7 13 15 17 22 23 13 18 19 19 14 11 18 21 8 11 20 20 18 23 16 17 23 23 10 19 18 16 16 21 11 12 15 23 0 16 9 18 13 18 21 23 19 16 15 13 21 0 23 20 20 15 18 10 7 18 10 22 15 22 11 16 22 20 17 16 20 18 15 18 7 16 3 15 19 21 2 5 10 19 13 0 3 1 21 19 8 8 20 2 21 14 20 21 10 22 21 22 14 20 14 13 10 15 17 23 12 11 23 1 20 8 11 7 14 16 7 20 6 5 5 6 7 17 7 20 1 14 15 6 21 21 21 1 11 20 5 17 21 21 23 2 0 9 13 10 0 9 15 10 13 4 13 23 9 11 15 23 11 9 23 22 18 21 21 20 16 13 14 8 20 12 12 11 7 20 9 23 18 7 22 18 21 11 9 20 23 20 21 10 9 12 0 0 22 23 17 16 23 22 11 12 11 8 19 18 23 15 19 2 22 7 13 15 17 19 17 16 1 19 22 20 20 22 21 16 9 19 1 9 3 0 2 1 18 20 15 23 14 9 13 18 11 0 8 7 19 13 14 18 9 17 6 19 20 13 15 1 12 6 11 17 22 11 21 21 12 22 5 19 22 20 19 20 9 23 18 21 23 10 18 13 0 0 22 8 1 1 18 20 6 19 8 22 5 5 13 14 23 10 21 12 21 1 0 10 17 1 15 17 9 22 18 20 11 16 19 22 13 23 19 22 12 19 11 16 18 4 13 17 16 17 12 22 11 19 17 17 14 18 11 22 13 12 22 9 11 23 12 9 17 21 3 14 16 11 21 3 0 12 5 22 2 20 15 15 19 12 19 8 19 16 13 14 21 21 22 23 2 13 13 6 9 19 17 18 19 9 7 3 3 19 17 18 9 20 17 1 1 2 7 18 15 7 7 20 3 19 21 12 4 12 23 1 18 23 13 0 20 1 20 2 20 0 20 22 1 4 2 12 18 10 15 20 16 8 22 22 0 7 22 19 21 22 21 1 15 0 21 2 14 4 22 22 22 22 1 1 1 22 1 18 1 22 23 19 20 14 16 19 14 16 16 0 1 19 16 15 23 23 13 22 11 4 18 12 3 1 23 19 11 19 19 12 16 17 14 21 1 1 3 3

Top 3 horas (Municipio de Guatemala):
  - 21:00 → 828 accidentes
  - 23:00 → 798 accidentes
  - 22:00 → 774 accidentes


Inciso 7 — Distribución por hora del día (Municipio de Guatemala)

Total de accidentes considerados: 13,003

Picos horarios

21:00 → 828 (≈ 6.4%)

23:00 → 798 (≈ 6.1%)

22:00 → 774 (≈ 6.0%)

20:00 → 768 (≈ 5.9%)

19:00 → 764 (≈ 5.9%)

La franja 19:00–22:00 concentra 3,134 accidentes (≈ 24.1% del total).

Horas con menor incidencia

04:00 → 284 (mínimo diario)

Franjas horarias (agregadas)

Madrugada [0–5]: 2,572 (≈ 19.8%)

Mañana [6–11]: 2,651 (≈ 20.4%)

Tarde [12–17]: 3,174 (≈ 24.4%)

Noche [18–23]: 4,606 (≈ 35.4%)

Lecturas clave

La noche (18–23) es la franja con mayor siniestralidad (más de un tercio de los casos), con un máximo entre 19:00 y 22:00.

La madrugada (0–5) concentra ~1 de cada 5 eventos, con el mínimo a las 04:00.

La incidencia aumenta desde las 17:00 y se mantiene elevada hasta cerca de medianoche.

